synapse/storage/chunk_ordered_table.py

```diff
@@ -38,6 +38,10 @@ class ChunkDBOrderedListStore(OrderedListStore):
     second chunk, then we say that the first chunk references the second, and
     thus forming a DAG.
 
+    The server may only have a subset of all events in a room, in which case
+    its possible for the server to have chunks that are unconnected from each
+    other. The ordering between unconnected chunks is arbitrary.
```

Comment:
I don't think there is anything stopping B being in the middle, since theoretically if you inserted the chunks in the order A, C, B, D, then that obeys all the constraints and so the KB algorithm will leave things alone (I think).